manpages/fics_addplayer.1
.\" -*- mode: nroff; -*-
.\"
.\" SPDX-FileCopyrightText: Copyright 2024 Markus Uhlin
.\" SPDX-License-Identifier: ISC
.\"
.Dd May 26, 2024
.Dt FICS_ADDPLAYER 1
.Os
.Sh NAME
.Nm fics_addplayer
.Nd creates a registered chess player
.Sh SYNOPSIS
.Nm fics_addplayer
.Bk -words
.Op Fl a
.Ek
.Aq UserName
.Aq FullName
.Aq EmailAddress
.Sh DESCRIPTION
.Nm
creates a registered chess player on a server that runs FICS.
An email will be sent to the given address with the login details.
.Sh OPTIONS
.Bl -tag -width Ds
.It Fl a
Make the player an administrator.
.El
.Sh EXAMPLES
.Bd -literal -offset indent
$ fics_addplayer nash "Richard V. Nash" nash@nomos.com
.Ed
.Sh SEE ALSO
.Xr fics 1 , Xr makerank 1


CHANGELOG.md
# Change Log #

All notable changes to this fork of FICS version 1.6.2 will be
documented in this file.

## [Unreleased] ##
- Added command-line option `l` (Display the legal notice and exit.)
- Replaced calls to `rand()` with `brand()` which uses
  `arc4random_uniform()`.

## [1.4.1] - 2024-05-26 ##
- **Added** command-line option `a` to the addplayer program. If given, it
  adds a player with admin privileges.
- **Added** usage of `time_t`.
- **Fixed** out of bounds array access in `match_command()`.
- **Fixed** resource leaks, i.e. missing calls to `fclose()`.
- **Fixed** usage of possibly uninitialized variables.

## [1.4] - 2024-05-20 ##
- **Added** usage of `time_t`. Multiple occurrences.
- **Changed** the program to create news index files even if no old ones
  are existent on the disk.
- **Fixed** clang warnings.
- **Fixed** sign compare (`-Wsign-compare`). Multiple occurrences.
- **Reformatted code** according to OpenBSD's KNF. Whole tree completed.
- **Replaced** unbounded string handling functions.

## [1.3] - 2024-05-05 ##
- **Added** parameter lists to many function declarations.
- **Added** usage of `ARRAY_SIZE()`.
- **Added** usage of `reallocarray()`. Multiple occurrences.
- **Added** usage of the functions from `err.h`.
- **Changed** the make install target to not overwrite the data messages
  in case they're already present.
- **Deleted** unused includes.
- **Fixed** passing argument 2 of `ReadGameAttrs()` from incompatible
  pointer type in `jsave_history()`.
- **Made** functions and variables private where possible.
- **Reformatted code** according to OpenBSD's KNF:
  - `algcheck.c`
  - `command.c`
  - `gamedb.c`
  - `movecheck.c`
  - `obsproc.c`
  - `talkproc.c`
  - ...
- **Replaced** unbounded string handling functions. _Multiple_
  occurrences.

## [1.2] - 2024-04-14 ##
- **Added** parameter lists to many function declarations.
- **Added** usage of `reallocarray()` which handles multiplication
  overflow.
- **Added** usage of `msnprintf()`, `mstrlcpy()` and `mstrlcat()` which
  detects and logs truncation.
- **Added** usage of the functions from `err.h`.
- **Checked out** the following files by tag 1.0, because the previous
  changes made to them possibly introduced game bugs:
  - `algcheck.c`
  - `algcheck.h`
  - `board.c`
  - `movecheck.c`
  - `movecheck.h`
- **Cleared** sensitive data with `explicit_bzero()`.
- **Fixed** an empty hostname in the addplayer program.
- **Fixed** bogus type for storing the return value of `time()`. It should
  really be `time_t`. Multiple occurrences.
- **Fixed** non-ANSI function declarations of functions. (Multiple
  occurrences.)
- **Made** functions and variables private where possible.
- **Reformatted code** according to OpenBSD's KNF:
  - `ratings.c`
  - `variable.c`
  - ...
- **Replaced** unbounded string handling functions. _Multiple_
  occurrences.

## [1.1] - 2024-03-30 ##
- **Added** `PRINTFLIKE()` and fixed many format errors.
- **Added** argument lists to many function declarations.
- **Added** initialization of many variables.
- **Added** newly written manual pages.
- **Added** usage of `ARRAY_SIZE()`
- **Changed** the addplayer program to generate passwords that are 8
  characters long (was 4).
- **Deleted** non-existent functions from the header files.
- **Deleted** obsolete and unused code
- **Deleted** unused includes.
- **Fixed** format strings that weren't string literals. Potentially
  insecure. (_Multiple_ occurrences.)
- **Fixed** incorrect buffer sizes
- **Fixed** unchecked return values
- **Made** functions and variables private (aka static) where possible.
- **Marked** functions that doesn't return `__dead`.
- **Marked** unused function parameters.
- **Redefined** `ASSERT()`
- **Reformatted code** according to OpenBSD's KNF.
- **Replaced** `rand()` calls with arc4random.
- **Replaced** _multiple_ `sprintf()` calls with `snprintf()` + truncation
  checks.
- **Replaced** _multiple_ occurrences of `strcpy()` and `strcat()` with
  size-bounded versions.
- **Switched to** the usage of the functions from `err.h` in multiple
  places for error handling.

## [1.0] - 2023-12-28 ##
- Added a new build system (and deleted the old).
- Added argument lists to many function declarations.
- Added auto-generation of the header file `ficspaths.h` and included
  it in `config.h`.
- Added better handling of memory allocation errors.
- Added usage of the `time_t` typedef in multiple places. This instead
  of `int`.
- Declared file-local functions and variables as `PRIVATE`.
- Deleted disabled code
- Did new revisions of the following files:
  - `board.c`
  - `eco.c`
  - `fics_addplayer.c`
  - `formula.c`
  - `makerank.c`
  - `network.c`
  - `shutdown.c`
  - `stdinclude.h`
  - `utils.c`
  - ...
- Fixed a `sscanf()` bug in `com_anews()`.
- Fixed a bug in `fix_time()` (did return a local address)
- Fixed bogus `crypt()` calls. (The second arg was wrong.)
- Fixed bugs in `process_move()`
- Fixed bugs in `stored_mail_moves()`
- Fixed comparison between pointer and integer in `com_inchannel()`.
- Fixed dead assignments
- Fixed implicit integers
- Fixed _multiple_ cases of use of possibly uninitialized variables.
- Fixed _multiple_ possible buffer overflows
- Fixed _multiple_ `sprintf()` format overflows
- Fixed the type of the variable passed to `strgtime()` and
  `strltime()`. Multiple occurrences.
- Fixed unused variables
- Reformatted code according to OpenBSD's KNF.
- Renamed functions in order to avoid conflicts with system
  declarations.

## 2023-12-07 ##
- Forked FICS version 1.6.2 made by Richard Nash.


FICS/fics_addplayer.c
/* fics_addplayer.c
 *
 */

/*
    fics - An internet chess server.
    Copyright (C) 1994  Richard V. Nash

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/

/* Revision history:
   name		email		yy/mm/dd	Change
   Richard Nash	              	94/03/07	Created
   Markus Uhlin			23/12/09	Revamped the file
   Markus Uhlin			23/12/23	Fixed crypt()
   Markus Uhlin			23/12/28	Switched to arc4random_uniform()
   Markus Uhlin			24/01/01	Size-bounded copying
   Markus Uhlin			24/04/01	Fixed empty hostname
   Markus Uhlin			24/04/04	Added usage of explicit_bzero()
   Markus Uhlin			24/05/25	Added command-line option 'a'
*/

#include "stdinclude.h"
#include "common.h"

#include <err.h>
#include <string.h>

#include "command.h"
#include "config.h"
#include "fics_getsalt.h"
#include "playerdb.h"
#include "utils.h"

#if __linux__
#include <bsd/string.h>
#endif

#define PASSLEN 8

PRIVATE char	*funame = NULL;
PRIVATE char	*fname = NULL;
PRIVATE char	*email = NULL;

PRIVATE int admin = 0;

PRIVATE void
add_handle_to_list(const char *handle)
{
	FILE	*fp;
	char	 path[1024];

	snprintf(path, sizeof path, "%s/admin", DEFAULT_LISTS);

	if ((fp = fopen(path, "a")) == NULL) {
		warn("%s: unable to open %s", __func__, path);
		return;
	}

	fprintf(fp, "%s\n", handle);
	fclose(fp);
}

PRIVATE __dead void
usage(char *progname)
{
	fprintf(stderr, "Usage: %s [-a] <UserName> <FullName> <EmailAddress>\n",
	    progname);
	exit(1);
}

PUBLIC int
main(int argc, char *argv[])
{
	char	 password[PASSLEN + 1];
	char	 salt[FICS_SALT_SIZE];
	char	 text[2048];
	int	 i;
	int	 p;

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {
			case 'a':
				admin = 1;
				break;
			case '?':
			case 'h':
			default:
				usage(argv[0]);
				break;
			}
		} else {
			if (funame == NULL)
				funame = argv[i];
			else if (fname == NULL)
				fname = argv[i];
			else if (email == NULL)
				email = argv[i];
			else
				usage(argv[0]);
		}
	}

	if (funame == NULL || fname == NULL || email == NULL)
		usage(argv[0]);

	/* Add the player here */
	if (strlen(funame) >= MAX_LOGIN_NAME)
		errx(1, "Player name is too long");
	else if (strlen(funame) < 3)
		errx(1, "Player name is too short");
	else if (!alphastring(funame)) {
		errx(1, "Illegal characters in player name. "
		    "Only A-Za-z allowed.");
	}

	player_init(0);
	p = player_new();

	if (!player_read(p, funame))
		errx(1, "%s already exists.", funame);

	parray[p].name		= xstrdup(funame);
	parray[p].login		= stolower(xstrdup(funame));
	parray[p].fullName	= xstrdup(fname);
	parray[p].emailAddress	= xstrdup(email);

	for (i = 0; i < PASSLEN; i++)
		password[i] = ('a' + arc4random_uniform(26));
	password[i] = '\0';

	if (strlcpy(salt, fics_getsalt(), sizeof salt) >= sizeof salt)
		errx(1, "salt truncated");

	parray[p].passwd	= xstrdup(crypt(password, salt));
	parray[p].registered	= 1;
	parray[p].rated		= 1;

	if (strcasecmp(parray[p].login, HADMINHANDLE) == 0 || admin) {
		parray[p].adminLevel = ADMIN_ADMIN;
		add_handle_to_list(parray[p].login);
	}

	player_save(p);

	printf("Added player account: >%s< >%s< >%s< >%s<\n",
	       funame, fname, email, password);
	printf("Admin: %s\n", (parray[p].adminLevel > 0 ? "Yes" : "No"));

	snprintf(text, sizeof text,
	    "\nYour player account has been created.\n\n"
	    "Login Name: %s\n"
	    "Full Name: %s\n"
	    "Email Address: %s\n"
	    "Initial Password: %s\n\n"

	    "If any of this information is incorrect, please\n"
	    "contact the administrator to get it corrected.\n\n"

	    "You may change your password with the password\n"
	    "command on the server.\n\n"

	    "Please be advised that if this is an unauthorized\n"
	    "duplicate account for you, by using it you take\n"
	    "the risk of being banned from accessing this chess\n"
	    "server.\n\n"

	    "To connect to the server and use this account:\n\n"
	    "\ttelnet %s 5000\n\n"
	    "and enter your handle name and password.\n\n"

	    "Regards,\n\nThe FICS admins\n", funame, fname, email, password,
	    SERVER_HOSTNAME);

	mail_string_to_address(email, "FICS Account Created", text);
	explicit_bzero(password, sizeof(password));
	explicit_bzero(text, sizeof(text));
	return 0;
}
